<a href="https://colab.research.google.com/github/emmanuelvillada573/laboratorio1/blob/main/taller_correlaci%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
from google.colab import drive
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
drive.mount('/content/drive')

df = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/DA_Talento/data/clientes_dataset.xlsx')

df.info()


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   edad              10 non-null     int64
 1   ingreso_mensual   10 non-null     int64
 2   visitas_web_mes   10 non-null     int64
 3   tiempo_sitio_min  10 non-null     int64
 4   paginas_vistas    10 non-null     int64
 5   clics_email       10 non-null     int64
 6   compras_previas   10 non-null     int64
 7   uso_descuento     10 non-null     int64
 8   satisfaccion      10 non-null     int64
 9   abandono_cliente  10 non-null     int64
dtypes: int64(10)
memory usage: 932.0 bytes


Responde:
1. ¿Cuántas variables tiene el dataset?
Respuesta: _10_____________________
2. ¿Cuál es la variable objetivo?
Respuesta: ____abandono_cliente__________________
3. ¿Qué tipo de variables observas? (numéricas / categóricas)
Respuesta:  #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   edad              10 non-null     int64
 1   ingreso_mensual   10 non-null     int64
 2   visitas_web_mes   10 non-null     int64
 3   tiempo_sitio_min  10 non-null     int64
 4   paginas_vistas    10 non-null     int64
 5   clics_email       10 non-null     int64
 6   compras_previas   10 non-null     int64
 7   uso_descuento     10 non-null     int64
 8   satisfaccion      10 non-null     int64
 9   abandono_cliente  10 non-null     int64______________________

In [7]:
corr = df.corr()
corr




,edad,ingreso_mensual,visitas_web_mes,tiempo_sitio_min,paginas_vistas,clics_email,compras_previas,uso_descuento,satisfaccion,abandono_cliente
edad,1.000000,0.985672,-0.938172,-0.816813,-0.886309,-0.791587,0.957013,-0.836004,0.954586,-0.794501
ingreso_mensual,0.985672,1.000000,-0.957109,-0.820961,-0.892808,-0.791304,0.967218,-0.886508,0.961636,-0.841432
visitas_web_mes,-0.938172,-0.957109,1.000000,0.877131,0.964320,0.907692,-0.958373,0.873556,-0.903068,0.779117
tiempo_sitio_min,-0.816813,-0.820961,0.877131,1.000000,0.940166,0.925110,-0.817029,0.770666,-0.770666,0.558069
paginas_vistas,-0.886309,-0.892808,0.964320,0.940166,1.000000,0.972924,-0.905679,0.809802,-0.841435,0.645311
clics_email,-0.791587,-0.791304,0.907692,0.925110,0.972924,1.000000,-0.811647,0.733213,-0.733213,0.541940
compras_previas,0.957013,0.967218,-0.958373,-0.817029,-0.905679,-0.811647,1.000000,-0.880705,0.880705,-0.720577
uso_descuento,-0.836004,-0.886508,0.873556,0.770666,0.809802,0.733213,-0.880705,1.000000,-0.791667,0.666667
satisfaccion,0.954586,0.961636,-0.903068,-0.770666,-0.841435,-0.733213,0.880705,-0.791667,1.000000,-0.875000
abandono_cliente,-0.794501,-0.841432,0.779117,0.558069,0.645311,0.541940,-0.720577,0.666667,-0.875000,1.000000


In [8]:
correlacion_abandono = corr['abandono_cliente'].sort_values(ascending=False)
display(correlacion_abandono)

,abandono_cliente
abandono_cliente,1.000000
visitas_web_mes,0.779117
uso_descuento,0.666667
paginas_vistas,0.645311
tiempo_sitio_min,0.558069
clics_email,0.541940
compras_previas,-0.720577
edad,-0.794501
ingreso_mensual,-0.841432
satisfaccion,-0.875000


Variable Correlación con abandono_cliente Tipo de relación (positiva/negativa)
edad ______ __x____
ingreso_mensual ______ ____x__
visitas_web_mes __x____ ______
tiempo_sitio_min __x____ ______
paginas_vistas ___x___ ______
clics_email _____x_ ______
compras_previas ______ __x____
uso_descuento ___x___ ______
satisfaccion ______ ___x___

Paso 4 — Selección de variables más relevantes

Las 3 variables con mayor correlación  con abandono_cliente son:

satisfaccion (-0.875)
ingreso_mensual (-0.841)
edad (-0.794)

La variable que tiene la correlación más alta (positiva o negativa) con abandono_cliente es satisfaccion con un valor de -0.875.

In [10]:
# Filtrar la variable objetivo 'abandono_cliente' para buscar correlaciones entre variables independientes
independent_corr = corr.drop(columns=['abandono_cliente']).drop(index=['abandono_cliente'])

# Encontrar pares de variables con correlación alta (valor absoluto > 0.9)
high_corr_pairs = []
for i in range(len(independent_corr.columns)):
    for j in range(i + 1, len(independent_corr.columns)):
        if abs(independent_corr.iloc[i, j]) > 0.9:
            pair = {
                'Variable 1': independent_corr.columns[i],
                'Variable 2': independent_corr.columns[j],
                'Correlación': independent_corr.iloc[i, j],
                '¿Son redundantes?': 'Sí' if abs(independent_corr.iloc[i, j]) > 0.9 else 'No'
            }
            high_corr_pairs.append(pair)

# Convertir a DataFrame para una mejor visualización
high_corr_df = pd.DataFrame(high_corr_pairs)

if not high_corr_df.empty:
    display(high_corr_df)
else:
    print("No se encontraron pares de variables independientes con una correlación (en valor absoluto) superior a 0.9.")

,Variable 1,Variable 2,Correlación,¿Son redundantes?
0,edad,ingreso_mensual,0.985672,Sí
1,edad,visitas_web_mes,-0.938172,Sí
2,edad,compras_previas,0.957013,Sí
3,edad,satisfaccion,0.954586,Sí
4,ingreso_mensual,visitas_web_mes,-0.957109,Sí
5,ingreso_mensual,compras_previas,0.967218,Sí
6,ingreso_mensual,satisfaccion,0.961636,Sí
7,visitas_web_mes,paginas_vistas,0.964320,Sí
8,visitas_web_mes,clics_email,0.907692,Sí
9,visitas_web_mes,compras_previas,-0.958373,Sí


### Paso 6 — Construcción del dataset reducido

**Variables seleccionadas (Top 5 con mayor correlación absoluta con `abandono_cliente`):**

*   `satisfaccion` (Correlación: -0.875)
*   `ingreso_mensual` (Correlación: -0.841)
*   `edad` (Correlación: -0.794)
*   `visitas_web_mes` (Correlación: 0.779)
*   `compras_previas` (Correlación: -0.721)

### PARTE 2 — Interpretación (Correlación de Pearson)

**Análisis conceptual**

**1. ¿Qué significa una correlación negativa en este contexto?**

En este contexto, una correlación negativa entre una variable independiente y `abandono_cliente` significa que a medida que el valor de la variable independiente aumenta, la probabilidad o tendencia de `abandono_cliente` disminuye. Por ejemplo, una correlación negativa fuerte entre `satisfaccion` y `abandono_cliente` indica que cuanto mayor es la satisfacción del cliente, menor es la probabilidad de que el cliente abandone.

**2. Explica la relación entre: satisfaccion y abandono_cliente**

La variable `satisfaccion` tiene una correlación de -0.875 con `abandono_cliente`. Esto es una correlación negativa muy fuerte. Significa que existe una relación inversa y casi lineal entre la satisfacción del cliente y el abandono. Específicamente, a medida que la satisfacción del cliente aumenta, la probabilidad de que un cliente abandone disminuye drásticamente. Esto sugiere que la satisfacción del cliente es un factor crucial para la retención y que mejorarla podría ser muy efectivo para reducir la tasa de abandono.

**3. Explica por qué una variable con correlación cercana a 0 no es útil.**

Una variable con una correlación cercana a 0 con la variable objetivo (en este caso, `abandono_cliente`) no es útil porque indica que no hay una relación lineal significativa entre ambas. Esto significa que los cambios en esa variable independiente no están linealmente asociados con los cambios en `abandono_cliente`. Para modelos lineales predictivos, esta variable aportaría muy poca o ninguna información para explicar la variabilidad de la variable objetivo, y por lo tanto, podría considerarse irrelevante para el modelo.

**4. ¿Por qué no debemos incluir variables altamente correlacionadas entre sí?**

No debemos incluir variables altamente correlacionadas entre sí (fenómeno conocido como **multicolinealidad**) en un modelo predictivo por varias razones:

*   **Redundancia:** Las variables altamente correlacionadas aportan información muy similar al modelo. Incluir ambas es redundante y no añade valor predictivo adicional significativo.
*   **Inestabilidad del modelo:** La multicolinealidad puede hacer que las estimaciones de los coeficientes de regresión sean inestables y difíciles de interpretar. Pequeños cambios en los datos pueden llevar a grandes cambios en los coeficientes.
*   **Dificultad de interpretación:** Es complicado determinar la contribución individual de cada variable a la predicción si están fuertemente relacionadas entre sí. Los errores estándar de los coeficientes tienden a ser altos, lo que hace que las variables parezcan no significativas incluso si en realidad lo son.
*   **Overfitting (sobreajuste):** En algunos casos, puede llevar a un modelo más complejo de lo necesario y propenso a sobreajustarse a los datos de entrenamiento.

Para evitar estos problemas, a menudo se selecciona solo una de las variables altamente correlacionadas o se utilizan técnicas de reducción de dimensionalidad.

Variable 1 Variable 2 Correlación ¿Son redundantes?:
	Variable 1	Variable 2	Correlación	¿Son redundantes?
0	edad	ingreso_mensual	0.985672	Sí
1	edad	visitas_web_mes	-0.938172	Sí
2	edad	compras_previas	0.957013	Sí
3	edad	satisfaccion	0.954586	Sí
4	ingreso_mensual	visitas_web_mes	-0.957109	Sí
5	ingreso_mensual	compras_previas	0.967218	Sí
6	ingreso_mensual	satisfaccion	0.961636	Sí
7	visitas_web_mes	paginas_vistas	0.964320	Sí
8	visitas_web_mes	clics_email	0.907692	Sí
9	visitas_web_mes	compras_previas	-0.958373	Sí
10	visitas_web_mes	satisfaccion	-0.903068	Sí
11	tiempo_sitio_min	paginas_vistas	0.940166	Sí
12	tiempo_sitio_min	clics_email	0.925110	Sí
13	paginas_vistas	clics_email	0.972924	Sí
14	paginas_vistas	compras_previas	-0.905679	Sí
